cpp/src/jni/types_helper.cc

```diff
@@ -0,0 +1,175 @@
+      gandiva_data_type->set_type(types::GandivaType::TIME32);
+      std::shared_ptr<arrow::Time32Type> cast_time_32_type =
+          std::dynamic_pointer_cast<arrow::Time32Type>(type);
+      arrow::TimeUnit::type unit = cast_time_32_type->unit();
+      types::TimeUnit time_unit = MapTimeUnit(unit);
+      gandiva_data_type->set_timeunit(time_unit);
+      break;
+    }
+    case arrow::Type::type::TIME64: {
+      gandiva_data_type->set_type(types::GandivaType::TIME32);
+      std::shared_ptr<arrow::Time64Type> cast_time_64_type =
+          std::dynamic_pointer_cast<arrow::Time64Type>(type);
+      arrow::TimeUnit::type unit = cast_time_64_type->unit();
+      types::TimeUnit time_unit = MapTimeUnit(unit);
+      gandiva_data_type->set_timeunit(time_unit);
+      break;
+    }
+    case arrow::Type::type::NA:
+      gandiva_data_type->set_type(types::GandivaType::NONE);
+      break;
+    case arrow::Type::type::MAP:
+    case arrow::Type::type::INTERVAL:
+    case arrow::Type::type::DECIMAL:
+    case arrow::Type::type::LIST:
+    case arrow::Type::type::STRUCT:
+    case arrow::Type::type::UNION:
+    case arrow::Type::type::DICTIONARY:
+      // un-supported types. test ensures that
+      // when one of these are added build breaks.
+      DCHECK(false);
+  }
+}
+
+JNIEXPORT jbyteArray JNICALL
+Java_org_apache_arrow_gandiva_evaluator_TypesJniHelper_getGandivaSupportedDataTypes(
+    JNIEnv *env, jobject types_helper) {
+  types::GandivaDataTypes gandiva_data_types;
+  auto supported_types = Types::supported_types();
+  for (auto const &type : supported_types) {
+    types::ExtGandivaType *gandiva_data_type = gandiva_data_types.add_datatype();
+    ArrowToProtobuf(type, gandiva_data_type);
+  }
+  size_t size = gandiva_data_types.ByteSizeLong();
+  jbyte buffer[size];
+  gandiva_data_types.SerializeToArray((void *)buffer, size);
+  jbyteArray ret = env->NewByteArray(size);
+  env->SetByteArrayRegion(ret, 0, size, buffer);
+  return ret;
+}
+
+/*
+ * Class:     org_apache_arrow_gandiva_types_TypesJNIHelper
+ * Method:    getGandivaSupportedFunctions
+ * Signature: ()[B
+ */
+JNIEXPORT jbyteArray JNICALL
+Java_org_apache_arrow_gandiva_evaluator_TypesJniHelper_getGandivaSupportedFunctions(
+    JNIEnv *env, jobject types_helper) {
+  auto supported_functions = Types::supported_functions();
+  types::GandivaFunctions gandiva_functions;
+  for (auto &function : supported_functions) {
+    types::FunctionSignature *function_signature = gandiva_functions.add_function();
+    function_signature->set_name(function->base_name());
+    types::ExtGandivaType *return_type = function_signature->mutable_returntype();
+    ArrowToProtobuf(function->ret_type(), return_type);
+    for (auto &param_type : function->param_types()) {
+      types::ExtGandivaType *proto_param_type = function_signature->add_paramtypes();
+      ArrowToProtobuf(param_type, proto_param_type);
+    }
+  }
+  size_t size = gandiva_functions.ByteSizeLong();
+  jbyte buffer[size];
```

Comment:
var-len arrays are discouraged in c++. didn't you get an error ?

https://stackoverflow.com/questions/1887097/why-arent-variable-length-arrays-part-of-the-c-standard

Reply:
it succeeded both in mac and travis. guess the new version of the compiler accepts it? will change it to dynamic otherwise..

Reply:
pls change it to dynamic. These usages can cause stack overflows.